### PR TITLE
simplify OS specific homedir detection

### PR DESCRIPTION
**Pull request by harshavardhana**

### CODE REVIEW

**Comment by stapelberg** [comment]
@mitchellh Ping?

**Comment by fd0** [comment]
It'd be nice to have this merged back in!

**Comment by harshavardhana** [comment]
Any way to get this merged ?

**Comment by mitchellh** [comment]
Hey @harshavardhana, thanks for the PR. Correct me if I'm wrong but could you explain the benefits of this PR? I see some code shuffling but curious on some decisions. For example, you changed the RWMutex to a Mutex which could maybe be better due to this usage but do you have reasons or data to back this up? 

You also added copyrights into every file which I don't do. Over the years of being deeply involved in OSS I've talked to no shortage of lawyers and the feedback on this goes back and forth but the lawyers I work with say that the copyright is not necessary since it is implicit based on the license and CLA (if there is one, this project does not have one). 

I'm fine with most of the changes and I can clean them up afterwards. Main question is the mutex changes probably!

**Comment by harshavardhana** [comment]
> For example, you changed the RWMutex to a Mutex which could maybe be better due to this usage but do you have reasons or data to back this up?

very historical @mitchellh perhaps i can revert it back. i am not really sure of the top of my head.

**Comment by harshavardhana** [comment]
@mitchellh there you go changed it back to RWMutex

File: homedir_test.go [hunk 1]
```diff
@@ -30,6 +30,8 @@ func BenchmarkDir(b *testing.B) {
 }
 
 func TestDir(t *testing.T) {
+	// NOTE: This test is not portable. If user.Current() worked
+	// everywhere, we wouldn't need our package in the first place.
```
Comment (renatoathaydes) [inline homedir_test.go]: hehe!

**Comment by harshavardhana** [comment]
@mitchellh Happy New Year! - would it be possible to get this merged if no more issues?

**Comment by wwboynton** [comment]
Is there any reason why this hasn't been merged yet, or couldn't be?

**Pull request closed without merging**